### ISSUE TITLE
More fields

### ISSUE THREAD

**Issue by tristanlatr** [twisted/pydoctor#365]
This issue tracks our advancement in the support of a broader set of fields. 

(almost) everything from the [epytext manual](http://epydoc.sourceforge.net/manual-fields.html) (fields section) should be applicable to pydoctor too. This issue focus on the finite set of fields

@mthuurne already said that the `@group:` field is not going to be supported. 

Here is the summary of the current pydoctor fields support and what fields would need to be supported to reach Epydoc + Sphinx like fields. Maybe not everything for the list is relevant ?

Fields working properly:

@param p: 
    A description of the parameter p for a function or method.
@type p: 
    The expected type for the parameter p.
@return: 
    The return value for a function or method.
@rtype: 
    The type of the return value for a function or method.
@keyword p: 
    A description of the keyword parameter p.
@raise e: 
    A description of the circumstances under which a function or method raises exception e. 
@ivar v: 
    A description of the class instance variable v.
@cvar v: 
    A description of the static class variable v.
@var v: 
    A description of the module variable v.
@type v: 
    The type of the variable v. 
@note: 
    A note about an object. Multiple note fields may be used to list separate notes.
@since: 
    The date or version when an object was first introduced.
@author: 
    The author(s) of an object. Multiple author fields may be used if an object has multiple authors.
@yields: (soon, will be used by pydoctor.napoleon)
@warns: (soon, will be used by pydoctor.napoleon)

Not supported yet:

@kwtype: Alias for @type:
@vartype: Alias for @type:
@param type p: 
   Idem as @param but with type specification inline (works only if single word types)
@attention: ...
    An important note about an object. Multiple attention fields may be used to list separate notes.
@bug: 
    A description of a bug in an object. Multiple bug fields may be used to report separate bugs.
@warning, and other admonitions: 
    A warning about an object. Multiple warning fields may be used to report separate warnings.
@version: 
    The current version of an object.
@todo [ver]: 
    A planned change to an object. If the optional argument ver is given, then it specifies the version for which the change will be made. Multiple todo fields may be used if multiple changes are planned.
@deprecated: 
    Indicates that an object is deprecated. The body of the field describe the reason why the object is deprecated.
@status: 
    The current status of an object.
@change: 
    A change log entry for this object.
@permission: 
    The object access permission, for systems such Zope/Plone supporting this concept. It may be used more than once to specify multiple permissions
@requires: 
    A requirement for using an object. Multiple requires fields may be used if an object has multiple requirements.
@precondition:
    A condition that must be true before an object is used. Multiple precondition fields may be used if an object has multiple preconditions.
@postcondition: 
    A condition that is guaranteed to be true after an object is used. Multiple postcondition fields may be used if an object has multiple postconditions.
@invariant: 
    A condition which should always be true for an object. Multiple invariant fields may be used if an object has multiple invariants. 
@organization: 
    The organization that created or maintains an object.
@copyright: 
    The copyright information for an object.
@license: 
    The licensing information for an object.
@contact: 
    Contact information for the author or maintainer of a module, class, function, or method. Multiple contact fields may be used if an object has multiple contacts. 

There is a bit of intersection with some of reST directives tho.. But is this an big issue ? We should try to have a shared implementation when possible. Like for `:deprecated:` -> `..deprecated::`, `:change:`, and also admonitions, `@warning`, etc.
Not 100% of implementation can be share because of issue #293, but we can make it very close.

**Comment by mthuurne** [comment]
I think there are two things to weigh here:
- how much code and documentation does a field require?
- is there any project using pydoctor or interested in using pydoctor that is either already using the field or interested in using it?

For example an alias takes little code to support, but if no-one is using the alias then we're just adding noise to our code and docs.

I'm not against the `group` field per se, but just parsing it is pointless if we don't use it in formatting the output. Doing something meaningful with that information means adding a significant amount of code and I'd rather only do that if I'm sure that there is a project that will use the feature.

For preconditions, postconditions and invariants, I'm not sure docstrings are the preferred medium. A library like `dpcontracts` allows you to add executable annotations, so you can verify that your conditions are satisfied during unit and system testing.

I'm not sure copyright and licensing information is useful to have in API docs, since they apply to using the implementation (source) and not to using the interface (what we're documenting). I think an `SPDX-License-Identifier` comment at the top of each module might be a more elegant way of communicating the license.

To summarize, I think we should look at each field individually and decide whether the use case is strong enough to make the effort to support it worthwhile.

**Comment by tristanlatr** [comment]
Ok thanks for the feedback, let me reduce the scope to a set of fields that, I my sens, should really be supported in pydoctor: 

- `@kwtype`: Alias for `@type`:
- `@vartype`: Alias for `@type`:
Both because it would allow to write Sphinx compatible object docstrings. Also I needed to change `pydoctor.napoleon` to use :type: everywhere to be understood by pydoctor, but that's not the best in my opinion. Ideally the output of `pydoctor.napoleon` should be compatible with sphinx, even if it's not a requirement. 

- Bibliographical fields (supported by docutils, but we override this): 
```
author
authors
contact
copyright
date
dedication
organization
revision
status
version
```

- `@license`

About the license information, etc, I think I should be supported, it's been seen in the wild quite often. The `sphinx.ext.napoleon` uses these for instance. 

- Admonitions
```
@attention
@warning
@note
@tip
and other
```

Admonitions as fields should be supported, because there is no way to do it with epytext other wise, They would render somewhat like the reST admonitions.

- Tracking change fields: 
```
@since:
@deprecated:
@change:
```
It is already has implemented in Epytext and it done in reST. 

- Automatic indexes of fields: I'm sure we could maintain tables by fields type/group of types in order to generate a "Index of changes". 

I think that's it!

- Here are the fields that I don't care about, and as you said, other tools offer the same kind of features:

`@permission`
`@requires`
`@precondition`
`@postcondition`
`@invariant`
`@organization`

It's all unnecessary to me.
